template: Add dataFile parameter property for including arbitrary text from files

mstokes-f5 opened the pull request:
Sometimes it is desirable to keep a portion of a template in a separate file and include it into the template text.
This can be done with parameters and the `dataFile` property:

```javascript
const fast = require('@f5devcentral/f5-fast-core');

const templatesPath = '/path/to/templatesdir'; // directory containing example.data
const dataProvider = new fast.FsDataProvider(templatesPath);
const yamldata = `
    definitions:
        var:
            dataFile: example
    template: |
        {{var}}
`;

fast.Template.loadYaml(yamldata, { dataProvider })
    .then((template) => {
        console.log(template.getParametersSchema());
        console.log(template.render({virtual_port: 443});
    });
```
The `FsDataProvider` will pick up on any files with the `.data` extension in the template set directory.
When referencing the file in a template, use the filename (without the extension) as a key.

Parameters with a `dataFile` property:

* are removed from `required`
* have their `default` set to the contents of the file
* given a default `format` of `hidden`

mstokes-f5 committed:
template: Add dataFile parameter property for including arbitrary text from files

mstokes-f5 committed:
template: Support base64-encoding data file contents

This is done with a new `base64` parameter property:

```yaml
definitions:
    var:
        dataFile: example
        base64: true
    template: |
        {{var}}
```

kaustriaf5 commented on lib/data_provider.js:
```diff
@@ -0,0 +1,128 @@
+/**
+ * DataProvider that fetches data from an atg-storage DataStore
+ */
+class DataStoreDataProvider {
```
Would it be better to create a "base" class DataProvider and have some classes extend it?

```
class BaseDataProvider {
    constructor(dataStore, cache) {
      this.storage = dataStore;
       this.cache = cache;
  }

  fetch(key) {
   return this.cache.fetch(key)
 }
...
...
 }
```

Also, the name seems to be for generic purpose but it has a property/verbiage specifically for template sets. Maybe we can do something like below?

```
class DataStoreDataProvider {
    constructor(datastore, datastoreKey) {
        this.dsKey = datastoreKey;
        super(datastore, new ResourceCache(
               dataName => dataStore.hasItem(this.dsKey)
...
...

   }
}
```

Moguri replied:
I have created a common `BaseDataProvider` class to address the first point. For the second point, all Providers currently operate on template sets.

mstokes-f5 replied:
I have created a common BaseDataProvider class to address the first point. For the second point, all Providers currently operate on template sets.

kaustriaf5 commented on lib/template_provider.js:
```diff
@@ -342,6 +370,47 @@ class FsTemplateProvider extends BaseTemplateProvider {
+            .then(() => dataFiles);
+    }
+
+    /**
+     * Get all data files known to the provider (op
+
```
Looks like copy-paste err

mstokes-f5 replied:
Yup

kaustriaf5 commented on lib/template_provider.js:
```diff
@@ -520,6 +589,39 @@ class DataStoreTemplateProvider extends BaseTemplateProvider {
+                return this.listSets();
+            })
+            .then(setNames => Promise.all(setNames.map(x => this.storage.getItem(x))))
+            .then(setData => setData.filter(x => x))
```
Looks like this is not needed here?

mstokes-f5 replied:
Looks like it; I'll remove it.

kaustriaf5 commented on lib/template.js:
```diff
@@ -387,6 +404,20 @@ class Template {
+                    }
+
+                    const dataFile = dataFiles[propDef.dataFile];
+                    if (propDef.base64) {
```
Would it be worth making this two way instead with two keywords?
 
`fromBase64` (which will decode before using the value) 
and `toBase64` (which is the current conversion)

mstokes-f5 replied:
I have added the two new keywords as suggested.

mstokes-f5 committed:
template: Rename `base64` property for dataFile to `toBase64` and add a `fromBase64`

mstokes-f5 committed:
DataProvider: Create a common BaseDataProvider class

mstokes-f5 committed:
SchemaProvider: Create a common BaseSchemaProvider class

mstokes-f5 commented:
In addition to creating a common `BaseDataProvider`, I went ahead and did `BaseSchemaProvider` as well. Now we have an abstract base class for all three provider types. I can move this change into a new PR if desired.

I am a little worried about adding a third provider type now that some other teams are starting to implement their own. It adds some complexity (see how I had to update the new GitHub providers that were added since this PR was first opened) and I don't know how many people actually need this feature. Luckily, both Schema and Data providers are pretty easy to implement and are a fair amount of copy/paste. This *does*, however, make me wonder if there is a better abstraction to combine providers that just need to get file contents (e.g., Schema and Data) vs loading Templates.

kaustriaf5 commented on lib/data_provider.js:
```diff
@@ -0,0 +1,150 @@
+    constructor(dataRootPath) {
+        super();
+
+        this.data_path = dataRootPath;
```
:nit: camelCase for consistency

mstokes-f5 replied:
Fixed. I also fixed up a similar issue with SchemaProvider (used an alias to maintain compatibility since the property is technically public).

mstokes-f5 committed:
DataProvider: Rename data_path to dataPath

mstokes-f5 committed:
SchemaProvider: Rename schema_path to schemaPath

Also include an alias so schema_path can still be used for backwards
compatibility.

kaustriaf5 commented:
Looks good to me, I had a nit comment that can be deferred..

kaustriaf5 commented:
> In addition to creating a common `BaseDataProvider`, I went ahead and did `BaseSchemaProvider` as well. Now we have an abstract base class for all three provider types. I can move this change into a new PR if desired.
> 
> I am a little worried about adding a third provider type now that some other teams are starting to implement their own. It adds some complexity (see how I had to update the new GitHub providers that were added since this PR was first opened) and I don't know how many people actually need this feature. Luckily, both Schema and Data providers are pretty easy to implement and are a fair amount of copy/paste. This _does_, however, make me wonder if there is a better abstraction to combine providers that just need to get file contents (e.g., Schema and Data) vs loading Templates.

Agreed that provider types seem to add some complexity and something we may want to review design for and revisit in the future, but for purposes of this MR/feature request, the changes look good, so I've approved.
